File: Leetcode/cpp/sol2409.cpp

// https://leetcode.com/problems/count-days-spent-together/description/

class Solution {
public:
    int ith_day(string date){
        vector <int> months = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
        string month = date.substr(0, 2);
        string day = date.substr(3, 2);
        int m = stoi(month);
        int d = stoi(day);
        int days = 0;
        for(int i = 0; i<m-1; i++){
            days+=months[i];
        }
        return days + d;
    }

    int countDaysTogether(string arriveAlice, string leaveAlice, string arriveBob, string leaveBob) {
        int alice_a = ith_day(arriveAlice);
        int alice_l = ith_day(leaveAlice);
        int bob_a = ith_day(arriveBob);
        int bob_l = ith_day(leaveBob);

        int start = max(alice_a, bob_a);
        int end = min(alice_l, bob_l);
        return max(0, end - start+1);

    }
};

File: Leetcode/python/sol2409.py
# https://leetcode.com/problems/count-days-spent-together/
class Solution:
    def ith_day(self, date:str) -> int:
        days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        month, day = date.split('-')
        return sum(days[:int(month)-1]) + int(day)

    def mark_dates(self, arrive:int, leave:int, dates:list) -> None:
        for i in range(arrive-1, leave):
            dates[i]+=1

    def countDaysTogether(self, arriveAlice: str, leaveAlice: str, arriveBob: str, leaveBob: str) -> int:
        alice_arrive  = self.ith_day(arriveAlice)
        alice_left = self.ith_day(leaveAlice)
        bob_arrive = self.ith_day(arriveBob)
        bob_left = self.ith_day(leaveBob)

        days = [0]*365 # passed by reference
        self.mark_dates(alice_arrive, alice_left, days)
        self.mark_dates(bob_arrive, bob_left, days)
        return sum(1 for i in days if i == 2)
